File: 06. Backtracking/04. Permutations I/main.cpp
#include <bits/stdc++.h>
using namespace std;

void backtrack(vector<int>& nums, vector<bool>& used, vector<int>& current, vector<vector<int>>& result) {
    if (current.size() == nums.size()) {
        result.push_back(current);
        return;
    }
    
    for (int i = 0; i < nums.size(); i++) {
        if (used[i]) continue; // Skip used numbers
        
        used[i] = true;
        current.push_back(nums[i]);
        backtrack(nums, used, current, result);
        current.pop_back();
        used[i] = false;
    }
}

vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> current;
    vector<bool> used(nums.size(), false);
    
    backtrack(nums, used, current, result);
    return result;
}

// Utility function to print permutations
void printPermutations(const vector<vector<int>>& permutations) {
    cout << "[\n";
    for (const auto& perm : permutations) {
        cout << "  [";
        for (size_t i = 0; i < perm.size(); i++) {
            cout << perm[i] << (i < perm.size() - 1 ? ", " : "");
        }
        cout << "]\n";
    }
    cout << "]\n";
}

int main() {
    vector<int> nums = {1, 2, 3};
    cout << "Permutations:\n";
    printPermutations(permute(nums));
    return 0;
}

File: 06. Backtracking/04. Permutations I/main.java
import java.util.*;

public class Permutations {
    public static void backtrack(List<List<Integer>> result, List<Integer> current, boolean[] used, int[] nums) {
        if (current.size() == nums.length) {
            result.add(new ArrayList<>(current));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            
            used[i] = true;
            current.add(nums[i]);
            backtrack(result, current, used, nums);
            current.remove(current.size() - 1);
            used[i] = false;
        }
    }

    public static List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, new ArrayList<>(), new boolean[nums.length], nums);
        return result;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3};
        System.out.println("Permutations: " + permute(nums));
    }
}
